tools/windows-node-installer/pkg/cloudprovider/azure/vm.go

```diff
@@ -933,6 +933,37 @@ func (az *AzureProvider) destroyDisk(ctx context.Context, vmInfo compute.Virtual
 	return
 }
 
+// DestroyWindowsVMWithResources destroys the Windows VMs along with all the errors associated with it
+func (az *AzureProvider) DestroyWindowsVM(vmName string) error {
+	ctx := context.Background()
+	_, nicName := az.getNICname(ctx, vmName)
+	_, ipName := az.getIPname(ctx, vmName)
+	vmInfo, _ := az.vmClient.Get(ctx, az.resourceGroupName, vmName, compute.InstanceView)
+
+	log.Printf("deleting the resources associated with instance %s", vmName)
+
+	err := az.destroyInstance(ctx, vmName)
+	if err != nil {
+		return fmt.Errorf("error deleting the instance %s: %v", vmName, err)
+	}
+
+	err = az.destroyNIC(ctx, nicName)
+	if err != nil {
+		return fmt.Errorf("error deleting the NIC of instance %s: %v", vmName, err)
```

Comment:
If this fails the other resources (IP/disk) will be left around and will be unfindable since we use the VM to find them.

Reply:
Not exactly, I think you'll still be able to find resources based on the VM name from console. The other approach is log and continue which I am not a fan either. I think if the api call fail, we should fail early instead of collecting all the failures and then reporting back(if that's what you're suggesting).

Reply:
I meant findable through our getNIC functions - but yeah not a great solution etiher way. Im fine with leaving it how it is

Reply:
I am nack on returning errors here. For the short term I prefer logging and ensuring that we don't leak resources. This becomes more important when we enable CI for Azure. Once we are closer to the product we can add these methods to as interface and explicitly destroy them from the Operator and use the same process for both AWS and Azure. But I don't see us getting there anytime soon.

Reply:
> I prefer logging and ensuring that we don't leak resources.

I made the changes to log and continue.
We'd still leak resources when the VM gets deleted and operator calls the `DestroyWindowsVM` method here -
https://github.com/openshift/windows-machine-config-bootstrapper/pull/159/files#diff-33ef6d4eee7a2e0183e61be55d8e3e79R967


For the record, I'd let the end-user deal with it. Even in our CI, if we properly create resources, the destroy should provide enough guarantees for VM deletion not the resources associated with it. Either we should we have external entity deal with it which is capable of deleting the lingering resources. If we start providing guarantees around the cloud provider api, we'll have to write a controller for wni :(.

Reply:
>We'd still leak resources when the VM gets deleted and operator calls the DestroyWindowsVM method here 

I don't follow this scenario. Why would the operator call the `DestroyWindowsVM()` method on a VM that has been deleted.

>For the record, I'd let the end-user deal with it. Even in our CI, if we properly create resources, the destroy should provide enough guarantees for VM deletion not the resources >associated with it. Either way we should have external entity deal with it which is capable of deleting the lingering resources. 

I agree with you and the point I am making is once we have this external entity capable of deleting lingering resources, we can change this code.

>If we start providing guarantees around the cloud >provider api, we'll have to write a controller for wni :(

I am not saying we will ever provide such guarantees.

Reply:
> I don't follow this scenario. Why would the operator call the DestroyWindowsVM() method on a VM that has been deleted.

If our node objects slice which has info of instanceIDs is stale, we can call the DestoryVM. Remember kubernetes is a eventually consistent system, so there will be stages where we get errors but we should be able to auto correct over time.